frontend/src/pages/MatchingGame.jsx

```diff
@@ -0,0 +1,118 @@
+                setMatchedPairs([...matchedPairs, selectedQuestion.question]);
+                setSelectedQuestion(null);
+                setSelectedAnswer(null);
+            } else {
+                setErrorMessage("Incorrect match! Try again.");
+            }
+        }
+    };
+
+    const isMatched = (item) =>
+        matchedPairs.includes(item.question || item.answer);
+
+    return (
+        <div className="bg-[#0f172a] min-h-screen flex flex-col items-center py-12 px-6 text-white">
+            <h1 className="text-4xl font-bold mb-8">Matching Game</h1>
+
+            <div className="grid grid-cols-4 gap-6 w-full max-w-4xl">
+                {/* Pytania */}
+                {shuffledQuestions.map((question, index) => (
+                    <div
+                        key={index}
+                        className={`p-4 rounded-lg cursor-pointer text-center bg-blue-600 ${
+                            isMatched(question) ? "opacity-50 pointer-events-none" : ""
+                        } ${
+                            selectedQuestion === question
+                                ? "ring-4 ring-yellow-500"
+                                : ""
+                        }`}
+                        onClick={() => handleQuestionClick(question)}
+                    >
+                        {question.question}
+                    </div>
+                ))}
+
+                {/* Odpowiedzi */}
+                {shuffledAnswers.map((answer, index) => (
+                    <div
+                        key={index}
+                        className={`p-4 rounded-lg cursor-pointer text-center bg-green-600 ${
+                            isMatched(answer) ? "opacity-50 pointer-events-none" : ""
+                        } ${
+                            selectedAnswer === answer
+                                ? "ring-4 ring-yellow-500"
+                                : ""
+                        }`}
+                        onClick={() => handleAnswerClick(answer)}
+                    >
+                        {answer.answer}
+                    </div>
+                ))}
+            </div>
+
+            {/* Komunikaty */}
+            <div className="mt-8 text-center">
+                {errorMessage && <p className="text-red-500 mb-4">{errorMessage}</p>}
+                <button
+                    className="form-button w-48"
+                    onClick={checkMatch}
+                    disabled={!selectedQuestion || !selectedAnswer}
+                >
+                    Check Match
+                </button>
+                <button
+                    className="form-button w-48 mt-4"
+                    onClick={() => {
+                        setSelectedTest(null);
+                        setMode(null);
+                    }}
+                >
+                    Back to My Tests
+                </button>
+            </div>
+
+            {/* Gratulacje po zakończeniu */}
```

Comment:
Let's change the comment to English.
```suggestion
            {/* Congratulation after ending */}
```